### ISSUE TITLE
Attached script executes incorrectly, repeatedly

### ISSUE THREAD

**Issue by captaindavepdx** [gildas-lormeau/SingleFile#201]
**Describe the bug**
The attached script(s) gets called repeatedly on dynamic websites - # depends on the website.

**To Reproduce**
1. Set the url in the command line to a dynamic website (airbnb, facebook, twitter, etc.)
2. Create a script to call an alert, one-time.
3. Add the script to the command line, "--browser-scripts"
4. Enable debug mode in the command line, "--browser-debug"
5. Run it.
6. See multiple alert messages - the number of alerts depends on the website.
7. Dismiss the alerts.
8. See more alerts.
9. Scroll down the page.
10. See more alerts.

**Expected behavior**
The attached script(s) should execute once, or according to the instructions in the script.

**Attached script: test.js**
`setTimeout(function(){alert("test");}, 2000)`

**Environment**
 - OS: MacOS
 - Browser: Chromium
 - Version: 75

**Additional context**
The attached screenshots are all from "airbnb.com"

**Screenshots**
<img width="449" alt="Screen Shot 2019-04-29 at 6 33 55 PM" src="https://user-images.githubusercontent.com/30560737/56936702-b6c8a500-6aad-11e9-9312-8658360f03f6.png">
<img width="446" alt="Screen Shot 2019-04-29 at 6 34 01 PM" src="https://user-images.githubusercontent.com/30560737/56936703-b6c8a500-6aad-11e9-9a88-14978a097a07.png">
<img width="448" alt="Screen Shot 2019-04-29 at 6 34 06 PM" src="https://user-images.githubusercontent.com/30560737/56936704-b6c8a500-6aad-11e9-9ed8-031a23b9f8a5.png">
<img width="448" alt="Screen Shot 2019-04-29 at 6 34 12 PM" src="https://user-images.githubusercontent.com/30560737/56936705-b6c8a500-6aad-11e9-9fb1-3e1e4849d459.png">
<img width="447" alt="Screen Shot 2019-04-29 at 6 34 18 PM" src="https://user-images.githubusercontent.com/30560737/56936706-b7613b80-6aad-11e9-81d3-c5aa88bd08cb.png">
<img width="442" alt="Screen Shot 2019-04-29 at 6 34 22 PM" src="https://user-images.githubusercontent.com/30560737/56936707-b7613b80-6aad-11e9-9805-4b8568eaab4d.png">
<img width="444" alt="Screen Shot 2019-04-29 at 6 34 28 PM" src="https://user-images.githubusercontent.com/30560737/56936708-b7613b80-6aad-11e9-9cdb-c70cb6be2e8f.png">
<img width="442" alt="Screen Shot 2019-04-29 at 6 34 34 PM" src="https://user-images.githubusercontent.com/30560737/56936709-b7613b80-6aad-11e9-8411-cb128818ebfb.png">
<img width="445" alt="Screen Shot 2019-04-29 at 6 34 39 PM" src="https://user-images.githubusercontent.com/30560737/56936710-b7613b80-6aad-11e9-9c1c-d5be96df3acf.png">
<img width="441" alt="Screen Shot 2019-04-29 at 6 34 45 PM" src="https://user-images.githubusercontent.com/30560737/56936711-b7f9d200-6aad-11e9-85d8-0b3accbc97bf.png">
<img width="441" alt="Screen Shot 2019-04-29 at 6 34 49 PM" src="https://user-images.githubusercontent.com/30560737/56936712-b7f9d200-6aad-11e9-8de6-04967d9632d8.png">
<img width="442" alt="Screen Shot 2019-04-29 at 6 34 53 PM" src="https://user-images.githubusercontent.com/30560737/56936713-b7f9d200-6aad-11e9-808d-378c600b7f47.png">
<img width="444" alt="Screen Shot 2019-04-29 at 6 34 58 PM" src="https://user-images.githubusercontent.com/30560737/56936714-b7f9d200-6aad-11e9-9bff-9b2b8b98b23c.png">
<img width="444" alt="Screen Shot 2019-04-29 at 6 35 03 PM" src="https://user-images.githubusercontent.com/30560737/56936715-b7f9d200-6aad-11e9-8eb9-f22c9b092104.png">
<img width="444" alt="Screen Shot 2019-04-29 at 6 35 11 PM" src="https://user-images.githubusercontent.com/30560737/56936716-b7f9d200-6aad-11e9-955b-981d4e7e159f.png">
<img width="443" alt="Screen Shot 2019-04-29 at 6 35 20 PM" src="https://user-images.githubusercontent.com/30560737/56936717-b7f9d200-6aad-11e9-8123-f9d68f09a438.png">
<img width="446" alt="Screen Shot 2019-04-29 at 6 35 26 PM" src="https://user-images.githubusercontent.com/30560737/56936718-b8926880-6aad-11e9-80e1-155ff97c8ca5.png">
<img width="445" alt="Screen Shot 2019-04-29 at 6 35 31 PM" src="https://user-images.githubusercontent.com/30560737/56936719-b8926880-6aad-11e9-90c9-d4654a29d516.png">
<img width="446" alt="Screen Shot 2019-04-29 at 6 35 36 PM" src="https://user-images.githubusercontent.com/30560737/56936720-b8926880-6aad-11e9-9541-7afa8a7f497a.png">
<img width="450" alt="Screen Shot 2019-04-29 at 6 35 40 PM" src="https://user-images.githubusercontent.com/30560737/56936721-b92aff00-6aad-11e9-9f08-daaf51d5d7db.png">
<img width="446" alt="Screen Shot 2019-04-29 at 6 33 34 PM" src="https://user-images.githubusercontent.com/30560737/56936722-b92aff00-6aad-11e9-9850-16373ba49bc4.png">
<img width="444" alt="Screen Shot 2019-04-29 at 6 33 43 PM" src="https://user-images.githubusercontent.com/30560737/56936723-b92aff00-6aad-11e9-8040-e57c85e23dd2.png">
<img width="449" alt="Screen Shot 2019-04-29 at 6 33 50 PM" src="https://user-images.githubusercontent.com/30560737/56936724-b92aff00-6aad-11e9-8c92-5ed423bd7db5.png">

**Comment by gildas-lormeau** [comment]
That's because the script is injected in all the frames. Run this instead `if (window == top) setTimeout(function(){alert("test");}, 2000);`.

**Comment by captaindavepdx** [comment]
Nice! Thanks, @gildas-lormeau!

We're so close, I was able to download the full height of a dynamic website - only a small layout issue

The winning combination appeared to have something to do with your "load-deferred-images" and "--load-deferred-images-max-idle-time,"  combined with my "auto-scroll" script, because it only worked when all three were added.

**Update:**
I did some script edits, and performed a variety of tests. The tests were performed in batches of three.

**Outcome:**
The script that failed 0/9 tests was triggered by the "load" event, after a 3000 ms delay, "setTimeout"

**Script triggered by the "DOMContentLoaded" event:** Failed 8/9 tests.
```
addEventListener("DOMContentLoaded", () => {

        // Scroll...       

    }, false);
```
**Script triggered by the "load" event:** Failed 5/9 tests.
```
addEventListener("load", () => {

        // Scroll...       

    }, false);
```
**Script triggered by the "DOMContentLoaded" event, after a 3000 ms delay:** Failed 1/9 tests.
```
addEventListener("DOMContentLoaded", () => {
        setTimeout(function() {

            // Scroll...   

        }, 3000)
    }, false);
```
**Script triggered by the "load" event, after a 3000 ms delay:** Failed 0/9 tests.
```
addEventListener("load", () => {
        setTimeout(function() {

            // Scroll...   

        }, 3000)
    }, false);
```

**Comment by captaindavepdx** [comment]
@gildas-lormeau Also, it's not a big deal, but the layout is off. Any ideas on what could be the cause?

**What it looks like:** If you drag down on the page, you can see the profile image and banner above.
<img width="1115" alt="Screen Shot 2019-04-30 at 11 22 01 AM" src="https://user-images.githubusercontent.com/30560737/56984148-39974180-6b3a-11e9-9d8d-9d27bee06152.png">

**What it should look like:**
<img width="1108" alt="Screen Shot 2019-04-30 at 11 22 09 AM" src="https://user-images.githubusercontent.com/30560737/56984149-39974180-6b3a-11e9-8009-35a49f56ec92.png">

**Comment by gildas-lormeau** [comment]
@captaindavepdx  Regarding the two issues. I need a precise scenario to know if it's the cause of SingleFile or not. I'm sorry but I won't troubleshoot the scraping of twitter pages for you. For example, in the second issue, I guess that the banner is not displayed *because* you scrolled. Is it hidden when you don't scroll?

**Comment by captaindavepdx** [comment]
@gildas-lormeau I'm confused, maybe I've miscommunicated? 

**In the first issue, the script was injected in all frames, which caused it to execute repeatedly. And that caused my script to work incorrectly.** 

You fixed that. 

Then, I posted an update. I let you know, the downloaded html file had improved, but it wasn't 100%.

Then, I posted my solution, and I included the steps to test and reproduce it.

**The "--load-deferred-images" and "--load-deferred-images-max-idle-time" didn't download the whole website by itself**
Even if I increased the idle time, the html file was either too short, or missing images, or both.

**My "scroll" script didn't download the whole website by itself either.**
It was too short, or missing images, or both.

**It only worked 100% when all three were added.**
And I had to edit my script to make it work correctly. I posted the script that worked 100%.

I want to help make SingleFile better, and I thought our work was doing that?

**Comment by captaindavepdx** [comment]
**In the second issue, the top of the website layout downloaded incorrectly.** 
I started to troubleshoot it, and posted the issue so we could troubleshoot it together, if you wanted to.

**Aren't those two issues, and previous issues, all related to SingleFile, and downloading websites?**
You added option "to execute scripts." That was awesome! And together, we have figured out how to: 
- Download websites using custom script(s) - more stable.
- Download websites that require login. 
- Download 100% of websites with dynamic content.

**Aren't those features that make SingleFile a little better? I enjoy troubleshooting with you, it's fun.**
To scrape Twitter, I'd use something like Casperjs, that can login and scrape data easily, out-of-the box.

**Comment by gildas-lormeau** [comment]
This is indeed a misunderstanding. The two issues I was referring to were your 2 previous posts before my answer.

**Comment by captaindavepdx** [comment]
**That's good, I'm glad we were able to correct the misunderstanding.**
I didn't think to post a final update in the previous posts, that's my bad. Instead, I opened a new post, and started troubleshooting the next bug. 

I can see how that could be confusing, and I'll make sure to be more thorough moving forward. 

Thankfully, we were able to fix the "--browser-height" and "twitter login" issues. 

**About the banner and avatar issue.**
Those sound like good ideas, thanks! I'll post an update after I try it.
